nowcasting_dataset/data_sources/sun/raw_data_load_save.py

```diff
@@ -0,0 +1,181 @@
+import datetime
+import time
+import io
+import logging
+from concurrent import futures
+from typing import List, Union, Optional
+
+import fsspec
+import numcodecs
+import pandas as pd
+from tqdm import tqdm
+import xarray as xr
+import numpy as np
+from pathlib import Path
+
+from nowcasting_dataset import geospatial
+
+logger = logging.getLogger(__name__)
+
+
+def get_azimuth_and_elevation(
+    datestamps: List[datetime.datetime], x_centers: List[int], y_centers: List[int]
+) -> (pd.DataFrame, pd.DataFrame):
+    """
+
+    Get Azimuth and elevation positions of the sun
+
+    For a list of datestamps and a list of coordinates, get the azimuth and elevation degrees.
+    Note that the degrees are rounded to 2 decimal places, as we at most need that.
+
+    Args:
+        datestamps: list of datestamps that are needed
+        x_centers: list of x coordinates - ref. OSGB
+        y_centers: list of y coordinates - ref. OSGB
+
+    Returns: Tuple of dataframes for azimuth and elevation.
+        The index is timestamps, and the columns are the x and y coordinates
+
+    """
+
+    logger.debug(
+        f"Will be calculating for {len(datestamps)} datestamps and {len(x_centers)} locations"
+    )
+
+    # create array of index datetime, columns of system_id for both azimuth and elevation
+    azimuth = []
+    elevation = []
+
+    t = time.time()
+    names = []
+    # loop over locations and find azimuth and elevation angles,
+    with futures.ThreadPoolExecutor() as executor:
```

Comment:
Nice!  I love the `futures` package too :)

if `geospatial.calculate_azimuth_and_elevation_angle()` is CPU-bound, then it may be a lot faster to use `future.ProcessPoolExecutor` because Python's threads all live in a single operating system thread.